src/Features/CSharp/Portable/RemoveAsyncModifier/CSharpRemoveAsyncModifierCodeFixProvider.cs

```diff
@@ -0,0 +1,106 @@
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.InitializeParameter;
+using Microsoft.CodeAnalysis.CSharp.MakeMethodSynchronous;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.RemoveAsyncModifier;
+using Roslyn.Utilities;
+using KnownTypes = Microsoft.CodeAnalysis.MakeMethodAsynchronous.AbstractMakeMethodAsynchronousCodeFixProvider.KnownTypes;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveAsyncModifier
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.RemoveAsyncModifier), Shared]
+    [ExtensionOrder(After = PredefinedCodeFixProviderNames.MakeMethodSynchronous)]
+    internal partial class CSharpRemoveAsyncModifierCodeFixProvider : AbstractRemoveAsyncModifierCodeFixProvider<ReturnStatementSyntax, ExpressionSyntax>
+    {
+        private const string CS1998 = nameof(CS1998); // This async method lacks 'await' operators and will run synchronously.
+
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public CSharpRemoveAsyncModifierCodeFixProvider()
+        {
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create(CS1998);
+
+        protected override bool IsAsyncSupportingFunctionSyntax(SyntaxNode node)
+            => node.IsAsyncSupportingFunctionSyntax();
+
+        protected override bool ShouldOfferFix(IMethodSymbol methodSymbol, KnownTypes knownTypes)
+        {
+            // For async void methods this fixer does the same as Make Method Synchronous so we don't need to offer both
+            if (methodSymbol.ReturnsVoid)
+            {
+                return false;
+            }
+
+            // IAsyncEnumerable iterators cannot be made non-async even if they don't have awaits
+            return !methodSymbol.ReturnType.OriginalDefinition.Equals(knownTypes._iAsyncEnumerableOfTTypeOpt);
+        }
+
+        protected override SyntaxNode GetLastChildOfBlock(SyntaxNode node)
+            => GetBlockBody(node).ChildNodes().Last();
+
+        protected override ControlFlowAnalysis AnalyzeControlFlow(SemanticModel semanticModel, SyntaxNode node)
+            => semanticModel.AnalyzeControlFlow(GetBlockBody(node));
+
+        private static SyntaxNode GetBlockBody(SyntaxNode node)
+            => node switch
+            {
+                MethodDeclarationSyntax method => method.Body,
+                LocalFunctionStatementSyntax localFunction => localFunction.Body,
+                AnonymousFunctionExpressionSyntax anonymousFunction => anonymousFunction.Block,
+                _ => throw ExceptionUtilities.Unreachable
+            };
+
+        protected override bool TryGetExpressionBody(SyntaxNode node, out SyntaxNode expression)
```

Comment:
consider using generics and type parameters on base type to more strongly type this to ExpressionSyntax (same applies to the Block code you have above).